examples/cds-hooks/cds-hooks-launch.js

```diff
@@ -0,0 +1,154 @@
+ *
+ * The EHR will call the cds-services route (the "discovery endpoint")
+ * in order to configure any available CDS services for this SMART application.
+ *
+ * Based on this configuration, the EHR may then post to /cds-services/patient-view
+ * with prefetch data as prescribed by the cds-services discovery route and,
+ * optionally, FHIR authorization details (i.e., access_token and scope). If
+ * an access token is provided, it may then be used by the FHIR client for
+ * requests to the EHR, as seen in the MedicationOrder example.
+
+ * Before every request, the EHR is first directed through the `authenticateEHR` middleware.
+ *
+ * `authenticateEHR` expects a JSON Web Token (JWT) from the EHR's authorization
+ * request header. It is used to establish that the request is from a trusted party.
+ * The JWT can be verified by one of 3 different ways in this example:
+ *
+ *  1) By setting a PEM file in the current directory on line 15.
+ *  2) By generating a PEM file from a `jku` variable set on line 16.
+ *  3) By generating a PEM file from a `jku` in the decoded JWT header.
+ *
+ */
+
+app.use(cors());
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(bodyParser.json());
+
+async function authenticateEHR(req, res, next) {
+  const token = req.headers.authorization.replace('Bearer ', '');
+  const decodedJwt = jwt.decode(token, { complete: true });
+  const asymmetricAlgs = ['ES256', 'ES384', 'ES384', 'RS256', 'RS384', 'RS512'];
+  const { alg, jku, kid } = decodedJwt.header;
+  const { iss, sub } = decodedJwt.payload;
+
+  let pem;
+  let verified;
+
+  const isWhitelisted = whitelistedEHRs.find(ehr => ehr.iss === iss && ehr.sub === sub);
+
+  if (!isWhitelisted) { return res.status(401).json('Authentication failed'); }
+
+  console.log(`token: ${token}`);
+  console.log(`decodedJwt: ${JSON.stringify(decodedJwt)}`);
+
+  if (asymmetricAlgs.includes(alg)) {
+    if (typeof pemPath !== 'undefined') {
+      // Use existing public key
+      try {
+        pem = fs.readFileSync(pemPath);
+      } catch (err) {
+        console.log(err);
+      }
+    } else if (typeof jku !== 'undefined') {
+      // Generate public key with an jwks.json endpoint
+      const jwks = await axios.get(jku);
+      const targetJwk = jwks.data.keys.find(key => key.kid === kid);
+
+      pem = jwkToPem(targetJwk);
+    }
+
+    try {
+      verified = jwt.verify(token, pem, { algorithms: [alg] });
+      console.log('Verified with PEM');
+    } catch (error) {
+      console.error('Invalid Token Error', error.message);
+      return res.status(401).json('Authentication failed');
+    }
+  }
+
+  console.log(`Authenticated Token: ${JSON.stringify(verified)}`);
+
+  return next();
+}
+
+async function authenticateClient(req, res, next) {
+  const { fhirServer, fhirAuthorization } = req.body;
+
+  if (typeof fhirAuthorization === 'undefined') {
+    return next();
+  }
+
+  req.fhirClient = new Client({ baseUrl: fhirServer });
```

Comment:
I still think this should be moved above the early return. The sandbox doesn't provide `fhirAuthorization`, so this example currently doesn't make any use of `fhir-kit-client` when using the cds-hooks sandbox. Using an unsecured FHIR server is intended for the cds-hooks sandbox http://cds-hooks.org/quickstart/#exposed-non-secured-fhir-server

http://cds-hooks.org/quickstart/#exposed-non-secured-fhir-server

Reply:
Does that launcher work for you? It doesn't for me.

Reply:
It does-- only thing I've run into is when using their CDS Hooks launch, the start of the ISS used in the URL from the launch needs to be manually changed from `http` to `https`. (Otherwise browsers block "loading mixed active content" from the unsecured ISS.)

Reply:
Ok, figured out my problem, I was using the STU3 server.